src/Microsoft.Health.Fhir.Liquid.Converter/Utilities/ComplexObjectFilterUtility.cs

```diff
@@ -0,0 +1,173 @@
+            var key = path.Dequeue();
+
+            // Check if our key is an object property or an array reference
+            if (key.Contains('['))
+            {
+                // Our required path dictates we should have an array and if not
+                // then this is not a match
+                if (input is object[] inputArray)
+                {
+                    // If key is [] then we want to loop over every object in the array
+                    // and check for our future paths
+                    if (key == "[]")
+                    {
+                        if (path.Count == 0)
+                        {
+                            return (value == null) || inputArray.Contains(value);
+                        }
+
+                        foreach (object obj in inputArray)
+                        {
+                            // Clone the queue to scope the path at this level for each loop
+                            var localPath = new Queue<string>(path);
+                            if (ObjHasValueAtPath(obj, localPath, value))
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+
+                    // If our array key references a specific index
+                    else
+                    {
+                        // Trim and parse out our numeric index
+                        var indexStr = key.TrimStart('[').TrimEnd(']');
+                        var index = int.Parse(indexStr);
+
+                        if (index >= 0 && inputArray.Count() > index)
+                        {
+                            // Recurse
+                            return ObjHasValueAtPath(inputArray[index], path, value);
+                        }
+                        else
+                        {
+                            // It's not possible for this array to have the element specified
+                            // so this is not a match
+                            return false;
+                        }
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            // Our key is referencing an object property
+            else
+            {
+                if (input is Dictionary<string, object> inputObject)
+                {
+                    if (inputObject.ContainsKey(key))
+                    {
+                        // Recurse
+                        return ObjHasValueAtPath(inputObject[key], path, value);
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        // This function splits a string into an array of parts like
+        // `test.path[].to[5].value` -> ['test', 'path', '[]', 'to', '[5]', 'value']
+        public static Queue<string> SplitObjectPath(string path)
```

Comment:
Maybe this didn't fit your use case but there is the `string.Split` methods that seem to do something similar. [Link](https://docs.microsoft.com/en-us/dotnet/api/system.string.split?view=netcore-3.1#system-string-split(system-char()-system-stringsplitoptions)). Seems like you could use it as a first call to split on the period characters and then loop through to find the brackets later.

Reply:
Good suggestion but I have a feeling this is going to be expanded upon with more DSL. Such as only checking a range or `<` `>` array indexes for example. In that future, I think we will want the ability to not split on a delimeter based on context, like I've done for the `.` here even though it doesn't get used yet.